### PR TITLE
add airflow + databricks guide

### PR DESCRIPTION
**Pull request by kentdanas**
@petedejoy @jwitz another airflow guide :) A few people from Databricks graciously reviewed this for us, so I've incorporated their feedback. Let me know if you have any comments!

### CODE REVIEW

**Comment by kentdanas** [comment]
not sure why that one link check is failing....it's just a post on the databricks forum and I can access it without any problems

File: guides/airflow-databricks.md [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+date: 2021-02-10T00:00:00.000Z
+slug: "airflow-databricks"
+tags: ["Integrations", "DAGs"]
+---
+
+> Note: All code in this guide can be found in [this Github repo](https://github.com/astronomer/airflow-databricks-tutorial).
+
+## Overview
+
+[Databricks](https://databricks.com/) is a popular unified data and analytics platform built around [Apache Spark](https://spark.apache.org/) that provides users with fully managed Apache Spark clusters and interactive workspaces. At Astronomer we believe that best practice is to use Airflow primarily as an orchestrator, and to leave the heavy lifting of data processing to an execution framework like Apache Spark. It follows that using the Airflow to orchestrate Databricks jobs is a natural solution for many common use cases.
```
Comment (jwitz) [inline guides/airflow-databricks.md:13]:
```suggestion
[Databricks](https://databricks.com/) is a popular unified data and analytics platform built around [Apache Spark](https://spark.apache.org/) that provides users with fully managed Apache Spark clusters and interactive workspaces. At Astronomer we believe that best practice is to use Airflow primarily as an orchestrator, and to use execution framework like Apache Spark to do the heavy lifting of data processing. It follows that using Airflow to orchestrate Databricks jobs is a natural solution for many common use cases.
```

File: guides/airflow-databricks.md [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+Astronomer has many customers that use Databricks to run jobs as part of complex pipelines. This can easily be accomplished by leveraging the [Databricks provider](https://github.com/apache/airflow/tree/master/airflow/providers/databricks), which includes Airflow hooks and operators that are actively maintained by the Databricks and Airflow communities. In this guide we will discuss the hooks and operators available to interact with Databricks clusters and run jobs, and show an example of how to use both available operators in an Airflow DAG.
```
Comment (jwitz) [inline guides/airflow-databricks.md:15]:
```suggestion
Astronomer has many customers who use Databricks to run jobs as part of complex pipelines. This can easily be accomplished by leveraging the [Databricks provider](https://github.com/apache/airflow/tree/master/airflow/providers/databricks), which includes Airflow hooks and operators that are actively maintained by the Databricks and Airflow communities. In this guide, we'll discuss the hooks and operators available for interacting with Databricks clusters and run jobs, and we'll show an example of how to use both available operators in an Airflow DAG.
```

File: guides/airflow-databricks.md [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+
+## Databricks Hooks and Operators
+
+The Databricks provider package includes many hooks and operators that allow users to accomplish most common Databricks-related use cases without writing a ton of code.
+
+### Hooks
+
+The [Databricks hook](https://github.com/apache/airflow/blob/master/airflow/providers/databricks/hooks/databricks.py) is the easiest and most efficient way to interact with a Databricks cluster or job from Airflow. The hook has methods to submit and run jobs to the Databricks REST API, which are used by the operators described below. There are also additional methods users can leverage to get information about runs or jobs, cancel, start, or terminate a cluster, and install and uninstall libraries on a cluster.
```
Comment (jwitz) [inline guides/airflow-databricks.md:23]:
```suggestion
The [Databricks hook](https://github.com/apache/airflow/blob/master/airflow/providers/databricks/hooks/databricks.py) is the most efficient way to interact with a Databricks cluster or job from Airflow. The hook has methods to submit and run jobs to the Databricks REST API, which are used by the operators described below. There are also additional methods users can leverage to get information about runs or jobs, cancel, start, or terminate a cluster, and install and uninstall libraries on a cluster.
```

Comment (jwitz) [reply guides/airflow-databricks.md:23]: Maybe break out the last sentence into some bullet points?

File: guides/airflow-databricks.md [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+
+### Operators
+
+There are currently two operators in the Databricks provider package: 
+
+- The `DatabricksSubmitRunOperator` makes use of the Databricks [Runs Submit API Endpoint](https://docs.databricks.com/dev-tools/api/latest/jobs.html#runs-submit) and submits a new Spark job run to Databricks.
+- The `DatabricksRunNowOperator` makes use of the Databricks [Run Now API Endpoint](https://docs.databricks.com/dev-tools/api/latest/jobs.html#run-now) and runs an existing Spark job.
+
+The `DatabricksRunNowOperator` should be used when you have an existing job defined in your Databricks workspace that you want to trigger using Airflow. The `DatabricksSubmitRunOperator` should be used if you want to manage the definition of your Databricks job and its cluster configuration within Airflow. Both operators allow you to run the job on a Databricks General Purpose cluster you have already created or on a separate Job Cluster that is created for the job and terminated upon the job’s completion.
```
Comment (jwitz) [inline guides/airflow-databricks.md:32]:
```suggestion
The `DatabricksRunNowOperator` should be used when you have an existing job defined in your Databricks workspace that you want to trigger using Airflow. The `DatabricksSubmitRunOperator` should be used if you want to manage the definition of your Databricks job and its cluster configuration within Airflow. Both operators allow you to run the job on a Databricks General Purpose cluster you've already created, or on a separate Job Cluster that is created for the job and terminated upon the job’s completion. 
```

File: guides/airflow-databricks.md [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+
+The `DatabricksRunNowOperator` should be used when you have an existing job defined in your Databricks workspace that you want to trigger using Airflow. The `DatabricksSubmitRunOperator` should be used if you want to manage the definition of your Databricks job and its cluster configuration within Airflow. Both operators allow you to run the job on a Databricks General Purpose cluster you have already created or on a separate Job Cluster that is created for the job and terminated upon the job’s completion. 
+
+
+Both operators are thoroughly documented in the [provider code](https://github.com/apache/airflow/blob/master/airflow/providers/databricks/operators/databricks.py); we recommend reading through the doc strings on both operators to get familiar with them.
+
+## Example - Using Airflow with Databricks
+
+Below we provide an example DAG that makes use of both the `DatabricksSubmitRunOperator` and the `DatabricksRunNowOperator`. Before diving into the DAG itself, we'll walk through a few prerequisites.
```
Comment (jwitz) [inline guides/airflow-databricks.md:39]:
```suggestion
Below, we provide an example DAG that makes use of both the `DatabricksSubmitRunOperator` and the `DatabricksRunNowOperator`. Before diving into the DAG itself, we'll walk through a few prerequisites.
```

File: guides/airflow-databricks.md [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+
+### Creating a Databricks Connection
+
+In order to use any Databricks hooks or operators, you will first need to create an Airflow connection that will allow Airflow to talk to your Databricks account. In general, Databricks recommends using a Personal Access Token (PAT) to authenticate to the Databricks REST API. For more information on how to generate a PAT for your account, check out the documentation [here](https://docs.databricks.com/dev-tools/data-pipelines.html).
```
Comment (jwitz) [inline guides/airflow-databricks.md:43]:
```suggestion
In order to use any Databricks hooks or operators, you will first need to create an Airflow connection that will allow Airflow to talk to your Databricks account. In general, Databricks recommends using a Personal Access Token (PAT) to authenticate to the Databricks REST API. For more information on how to generate a PAT for your account, read the [Databricks documentation](https://docs.databricks.com/dev-tools/data-pipelines.html). 
```

File: guides/airflow-databricks.md [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+
+## Example - Using Airflow with Databricks
+
+Below we provide an example DAG that makes use of both the `DatabricksSubmitRunOperator` and the `DatabricksRunNowOperator`. Before diving into the DAG itself, we'll walk through a few prerequisites.
+
+### Creating a Databricks Connection
+
+In order to use any Databricks hooks or operators, you will first need to create an Airflow connection that will allow Airflow to talk to your Databricks account. In general, Databricks recommends using a Personal Access Token (PAT) to authenticate to the Databricks REST API. For more information on how to generate a PAT for your account, check out the documentation [here](https://docs.databricks.com/dev-tools/data-pipelines.html). 
+
+For this example we have used the PAT authentication method, and have set up a connection using the Airflow UI. It should look something like this:
```
Comment (jwitz) [inline guides/airflow-databricks.md:45]:
```suggestion
For this example, we use the PAT authentication method and set up a connection using the Airflow UI. It should look something like this:
```

File: guides/airflow-databricks.md [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+
+![Databricks Connection](https://assets2.astronomer.io/main/guides/databricks-tutorial/databricks_connection.png)
+
+The Host should be your Databricks workspace URL, and your PAT should be added as a JSON-formatted Extra, as shown in the screenshot.
```
Comment (jwitz) [inline guides/airflow-databricks.md:49]:
```suggestion
The Host should be your Databricks workspace URL, and your PAT should be added as a JSON-formatted Extra.
```

File: guides/airflow-databricks.md [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+
+Note that it is also possible to use your login credentials to authenticate, although this isn't Databricks' recommended method of authentication. To use this method, you would enter the username and password you use to sign in to your Databricks account in the Login and Password fields of the connection.
+
+### Creating a Databricks Job
+
+In order to use the `DatabricksRunNowOperator` you must have a job already defined in your Databricks workspace. If you are new to creating jobs on Databricks, [this guide](https://docs.databricks.com/jobs.html) walks through all the basics.
+
+To follow the example DAG below, you will want to create a job that has a cluster attached, and has a parameterized notebook as a task. For more information on parameterizing a notebook, see [this page](https://docs.databricks.com/notebooks/widgets.html).
```
Comment (jwitz) [inline guides/airflow-databricks.md:57]:
```suggestion
To follow the example DAG below, you will want to create a job that has a cluster attached and a parameterized notebook as a task. For more information on parameterizing a notebook, see [this page](https://docs.databricks.com/notebooks/widgets.html).
```

File: guides/airflow-databricks.md [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+
+Once you have created a job, you should be able to see it in the Databricks UI Jobs tab like this:
```
Comment (jwitz) [inline guides/airflow-databricks.md:59]:
```suggestion
Once you have created a job, you should be able to see it in the Databricks UI Jobs tab:
```

File: guides/airflow-databricks.md [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+        job_id=5,
+        notebook_params=notebook_params
+    )
+
+    opr_submit_run >> opr_run_now
+```
+
+For both operators we need to provide the `databricks_conn_id` and necessary parameters. 
+
+For the `DatabricksSubmitRunOperator`, we need to provide parameters for the cluster that will be spun up (`new_cluster`). This should include, at a minimum, the spark version, number of workers, and node type ID, but can be defined more granularly as needed. For more information on what Spark version runtimes are available, see the documentation [here](https://docs.databricks.com/dev-tools/api/latest/index.html#runtime-version-strings).
```
Comment (jwitz) [inline guides/airflow-databricks.md:122]:
```suggestion
For the `DatabricksSubmitRunOperator`, we need to provide parameters for the cluster that will be spun up (`new_cluster`). This should include, at a minimum: 

* Spark version
* Number of workers
* Node type ID 

These can be defined more granularly as needed. For more information on what Spark version runtimes are available, see the [Databricks documentation](https://docs.databricks.com/dev-tools/api/latest/index.html#runtime-version-strings).
```

File: guides/airflow-databricks.md [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+
+We also need to provide the task that will be run. In this example we provide the `notebook_task`, which is the path to the Databricks notebook we want to run. Note that this could alternatively be a Spark JAR task, Spark Python task, or Spark submit task, which would be defined using the `spark_jar_task`, `spark_python_test`, or `spark_submit_task` parameters respectively. The operator will look for one of these four options to be defined.
+
+For the `DatabricksRunNowOperator`, we only need to provide the `job_id` for the job we want to submit, since the job parameters should already be configured in Databricks. You can find on the `job_id` on the Jobs tab of your Databricks account as shown in the screenshot above. However, you can also provide `notebook_params`, `python_params` or `spark_submit_params` as needed for your job. In this case we have parameterized our notebook to take in a `Variable` integer parameter, and have passed in '5' for this example.
```
Comment (jwitz) [inline guides/airflow-databricks.md:126]:
```suggestion
For the `DatabricksRunNowOperator`, we only need to provide the `job_id` for the job we want to submit, since the job parameters should already be configured in Databricks. You can find on the `job_id` on the Jobs tab of your Databricks account as shown in the screenshot above. However, you can also provide `notebook_params`, `python_params`, or `spark_submit_params` as needed for your job. In this case, we parameterized our notebook to take in a `Variable` integer parameter and passed in '5' for this example.
```

File: guides/airflow-databricks.md [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+
+### Error Handling
+
+When using either of these operators, any failures in submitting the job, starting or accessing the cluster, or connecting with the Databricks API, will propagate to a failure of the Airflow task and error messages will be shown in the logs.
```
Comment (jwitz) [inline guides/airflow-databricks.md:130]:
```suggestion
When using either of these operators, any failures in submitting the job, starting or accessing the cluster, or connecting with the Databricks API will propagate to a failure of the Airflow task and generate an error message in the logs.
```

File: guides/airflow-databricks.md [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+
+If there is a failure in the job itself, like in one of the notebooks in this example, that failure will also propagate to a failure of the Airflow task. In that case the error message may not be shown in the Airflow logs, but the logs should include a URL link to the Databricks job status which will include errors, print statements, etc. For example, if we set up the notebook in Job ID 5 in the example above to have a bug in it, we get a failure in the task and the Airflow task log looks something like this:
```
Comment (jwitz) [inline guides/airflow-databricks.md:132]:
```suggestion
If there is a failure in the job itself, like in one of the notebooks in this example, that failure will also propagate to a failure of the Airflow task. In that case, the error message may not be shown in the Airflow logs, but the logs should include a URL link to the Databricks job status which will include errors, print statements, etc. For example, if we set up the notebook in Job ID 5 in the example above to have a bug in it, we get a failure in the task causing the Airflow task log to look something like this:
```

File: guides/airflow-databricks.md [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+
+![Error Log Example](https://assets2.astronomer.io/main/guides/databricks-tutorial/databricks_failure_airflow_log.png)
+
+In the case above we could click on the URL link to get to the Databricks log in order to debug the issue.
```
Comment (jwitz) [inline guides/airflow-databricks.md:136]:
```suggestion
In the case above, we could click on the URL link to get to the Databricks log in order to debug the issue.
```

File: guides/airflow-databricks.md [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+
+## Where to Go From Here
+
+This example DAG shows how little code is required to get started orchestrating Databricks jobs with Airflow. By using existing hooks and operators, you can easily manage your Databricks jobs from one place, while also building in any other steps in your data pipelines. With just a few more tasks, you can turn the DAG above into a pipeline for orchestrating many different systems like this:
```
Comment (jwitz) [inline guides/airflow-databricks.md:140]:
```suggestion
This example DAG shows how little code is required to get started orchestrating Databricks jobs with Airflow. By using existing hooks and operators, you can easily manage your Databricks jobs from one place while also building your data pipelines. With just a few more tasks, you can turn the DAG above into a pipeline for orchestrating many different systems like this:
```

**Review by jwitz — COMMENTED** [review]
Minor comments and suggestions throughout. Hats off to @kentdanas !

File: guides/airflow-databricks.md [hunk 17]
```diff
@@ -0,0 +1,154 @@
...
+
+## Overview
+
+[Databricks](https://databricks.com/) is a popular unified data and analytics platform built around [Apache Spark](https://spark.apache.org/) that provides users with fully managed Apache Spark clusters and interactive workspaces. At Astronomer we believe that best practice is to use Airflow primarily as an orchestrator, and to use an execution framework like Apache Spark to do the heavy lifting of data processing. It follows that using Airflow to orchestrate Databricks jobs is a natural solution for many common use cases.
+
+Astronomer has many customers who use Databricks to run jobs as part of complex pipelines. This can easily be accomplished by leveraging the [Databricks provider](https://github.com/apache/airflow/tree/master/airflow/providers/databricks), which includes Airflow hooks and operators that are actively maintained by the Databricks and Airflow communities. In this guide, we'll discuss the hooks and operators available for interacting with Databricks clusters and run jobs, and we'll show an example of how to use both available operators in an Airflow DAG.
```
Comment (mik-laj) [inline guides/airflow-databricks.md:15]: why did you add links to the source code instead of documentation?

Comment (petedejoy) [reply guides/airflow-databricks.md:15]: Don't think it was intentional- some folks just like being able to browse the source.

**Review by petedejoy — APPROVED** [review]
LGTM- great work @kentdanas!
